--- /dev/null	2010-07-18 22:35:57.354852321 -0400
+++ alix-madwifi-src/package/base-files/files/etc/init.d/x86setup	2010-07-24 15:53:37.734279654 -0400
@@ -0,0 +1,36 @@
+#!/bin/sh /etc/rc.common
+
+START=15
+
+boot()
+{
+	eths=$(cat /proc/net/dev | grep eth | sed 's/^.*eth/eth/g' | sed 's/:.*$//g')
+	num_eths=$(echo $eths | wc -w)
+		
+	wan_if_num=$(($num_eths-1))
+	lan_ifs=""
+	if_num=0
+	for eth in $eths ; do
+		if [ $if_num -gt 0 ] && [ $if_num = $wan_if_num ] ; then
+			uci set network.wan="interface"
+			uci set network.wan.ifname="$eth"
+			uci set network.wan.proto="dhcp"
+		else
+			if [ $if_num = 0 ] ; then
+				lan_ifs="$eth"
+			else
+				lan_ifs="$lan_ifs $eth"
+			fi
+		fi
+		if_num=$(($if_num+1))
+	done
+	
+	if [ -n "$lan_ifs" ] ; then
+		uci set network.lan="interface"
+		uci set network.lan.ifname="$lan_ifs"
+		uci set network.lan.type="bridge"
+	fi
+	uci commit
+	/etc/init.d/x86setup disable
+	rm -rf /etc/init.d/x86setup
+}
--- backfire-src/package/base-files/Makefile	2010-06-29 23:25:45.114195070 -0400
+++ alix-madwifi-src/package/base-files/Makefile	2010-07-24 16:49:01.934279633 -0400
@@ -386,6 +386,8 @@
 				echo "$$$${conffile##$(1)}" >> $(1)/CONTROL/conffiles; \
 		fi \
 	done
+	chmod 755 $(1)/etc/init.d/*
+
 endef
 
 ifeq ($(CONFIG_EXTERNAL_TOOLCHAIN),)
