--- kamikaze-7.09-src/target/sdk/files/Makefile	2007-09-20 04:54:28.000000000 -0400
+++ kamikaze-7.09-gargoyle-src/target/sdk/files/Makefile	2008-11-29 22:01:38.000000000 -0500
@@ -74,9 +74,9 @@
 	$(MAKE) package/compile
 	-( \
 		cd package; \
-		find . -maxdepth 2 -name Config.in | \
-			sed -e 's,/Config.in,,g' | \
-			xargs -n1 $(MAKE) compile -C; \
+		for configfile in `find . -maxdepth 2 -name Config.in` ; do \
+			$(MAKE) compile -C `dirname $$configfile` ; \
+		done \
 	)
 
 clean: FORCE
