--- kamikaze-7.09-src/package/broadcom-wl/files/lib/wifi/broadcom.sh	2007-09-20 04:54:54.000000000 -0400
+++ kamikaze-7.09-gargoyle-src/package/broadcom-wl/files/lib/wifi/broadcom.sh	2008-11-29 22:01:38.000000000 -0500
@@ -52,7 +52,7 @@
 	case "$adhoc:$sta:$apmode" in
 		1*)
 			ap=0
-			mssid=0
+			mssid=
 			infra=0
 		;;
 		:1:1)
@@ -62,7 +62,7 @@
 		:1:)
 			wet=1
 			ap=0
-			mssid=0
+			mssid=
 		;;
 		::)
 			radio=0
@@ -97,6 +97,9 @@
 	config_get slottime "$device" slottime
 	config_get rxant "$device" rxant
 	config_get txant "$device" txant
+	config_get_bool frameburst "$device" frameburst
+	config_get macfilter "$device" macfilter
+	config_get maclist "$device" maclist
 	local vif_pre_up vif_post_up vif_do_up
 
 	_c=0
@@ -112,6 +115,18 @@
 	} || {
 		slottime="${slottime:--1}"
 	}
+	
+	case "$macfilter" in
+		allow|2)
+			macfilter=2;
+		;;
+		deny|1)
+			macfilter=1;
+		;;
+		disable|none|0)
+			macfilter=0;
+		;;
+	esac
 
 	for vif in $vifs; do
 		config_get mode "$vif" mode
@@ -144,11 +159,11 @@
 							config_get k "$vif" key$knr
 							[ -n "$k" ] || continue
 							[ "$defkey" = "$knr" ] && def="=" || def=""
-							append vif_pre_up "wepkey $def$knr,$k" "$N"
+							append vif_do_up "wepkey $def$knr,$k" "$N"
 						done
 					;;
 					"");;
-					*) append vif_pre_up "wepkey =1,$key" "$N";;
+					*) append vif_do_up "wepkey =1,$key" "$N";;
 				esac
 			;;
 			*psk*|*PSK*)
@@ -177,17 +192,15 @@
 				nasopts="-r \"\$${vif}_key\" -h $server -p $port"
 			;;
 		esac
-		append vif_post_up "wsec $wsec" "$N"
-		append vif_post_up "wpa_auth $auth" "$N"
-		append vif_post_up "wsec_restrict $wsec_r" "$N"
-		append vif_post_up "eap_restrict $eap_r" "$N"
+		append vif_do_up "wsec $wsec" "$N"
+		append vif_do_up "wpa_auth $auth" "$N"
+		append vif_do_up "wsec_restrict $wsec_r" "$N"
+		append vif_do_up "eap_restrict $eap_r" "$N"
 		
 		config_get ssid "$vif" ssid
 		append vif_post_up "vlan_mode 0" "$N"
 		append vif_post_up "ssid $ssid" "$N"
-		case "$mode" in
-			sta|adhoc) append vif_do_up "ssid $ssid" "$N";;
-		esac
+		append vif_do_up "ssid $ssid" "$N"
 		
 		append vif_post_up "enabled 1" "$N"
 		
@@ -207,8 +220,8 @@
 			[ "$mode" = "sta" ] && {
 				nas_mode="-S"
 				[ -z "$bridge" ] || {
-					append vif_pre_up "supplicant 1" "$N"
-					append vif_pre_up "passphrase $key" "$N"
+					append vif_post_up "supplicant 1" "$N"
+					append vif_post_up "passphrase $key" "$N"
 					
 					use_nas=0
 				}
@@ -221,9 +234,9 @@
 	wlc stdin <<EOF
 $ifdown
 
-ap $ap
-mssid $mssid
 apsta $apsta
+ap $ap
+${mssid:+mssid $mssid}
 infra $infra
 ${wet:+wet 1}
 802.11d 0
@@ -232,14 +245,15 @@
 txant ${txant:-3}
 
 radio ${radio:-1}
-macfilter 0
-maclist none
+macfilter ${macfilter:-0}
+maclist ${maclist:-none}
 wds none
 ${wds:+wds $wds}
-${channel:+channel $channel}
 country ${country:-IL0}
+${channel:+channel $channel}
 maxassoc ${maxassoc:-128}
 slottime ${slottime:--1}
+${frameburst:+frameburst $frameburst}
 
 $vif_pre_up
 up
