diff --git a/package/gargoyle/files/www/basic.sh b/package/gargoyle/files/www/basic.sh
index 0a69453c..75f613ed 100755
--- a/package/gargoyle/files/www/basic.sh
+++ b/package/gargoyle/files/www/basic.sh
@@ -351,9 +363,7 @@ var isb43 = wirelessDriver == "mac80211" && (!GwifiN) ? true : false ;
 				<div id="bridge_pass_container" class="row form-group">
 					<label class="col-xs-5" for="bridge_pass" id="bridge_pass_label"><%~ Pswd %>:</label>
 					<span class="col-xs-7">
-						<input type="password" id="bridge_pass" class="form-control" size="20" onkeyup="proofreadLengthRange(this,8,999)"/>
-						<input type="checkbox" id="show_bridge_pass" onclick="togglePass('bridge_pass')" />
-						<label for="show_bridge_pass" id="show_bridge_pass_label"><%~ rvel %></label>
+						<input type="text" id="bridge_pass" class="form-control" size="20" onkeyup="proofreadLengthRange(this,8,999)"/>
 					</span>
 				</div>
 
@@ -434,7 +456,7 @@ var isb43 = wirelessDriver == "mac80211" && (!GwifiN) ? true : false ;
 
 				<div id="wan_pppoe_pass_container" class="row form-group">
 					<label class="col-xs-5" for="wan_pppoe_pass" id="wan_pppoe_pass_label"><%~ Pswd %>:</label>
-					<span class="col-xs-7"><input type="password" class="form-control" id="wan_pppoe_pass" size="20" onkeyup="proofreadLengthRange(this,1,999)"/></span>
+					<span class="col-xs-7"><input type="text" class="form-control" id="wan_pppoe_pass" size="20" onkeyup="proofreadLengthRange(this,1,999)"/></span>
 				</div>
 
 				<div id="wan_pppoe_reconnect_mode_container" class="row form-group">
@@ -853,9 +875,7 @@ var isb43 = wirelessDriver == "mac80211" && (!GwifiN) ? true : false ;
 				<div id="wifi_pass2_container" class="row indent">
 					<label class="col-xs-5" for="wifi_pass2" id="wifi_pass2_label"><%~ Pswd %>:</label>
 					<span class="col-xs-7">
-						<input type="password" id="wifi_pass2" class="form-control" size="20" onkeyup="proofreadLengthRange(this,8,999)"/>&nbsp;&nbsp;
-						<input type="checkbox" id="show_pass2" onclick="togglePass('wifi_pass2')"/>
-						<label for="show_pass2" id="show_pass2_label"><%~ rvel %></label><br/>
+						<input type="text" id="wifi_pass2" class="form-control" size="20" onkeyup="proofreadLengthRange(this,8,999)"/>&nbsp;&nbsp;
 					</span>
 				</div>
 
@@ -934,9 +954,7 @@ var isb43 = wirelessDriver == "mac80211" && (!GwifiN) ? true : false ;
 				<div id="wifi_pass1_container" class="row indent">
 					<label class="col-xs-5" for="wifi_pass1" id="wifi_pass1_label"><%~ Pswd %>:</label>
 					<span class="col-xs-7" >
-						<input type="password" id="wifi_pass1" class="form-control" size="20" onkeyup="proofreadLengthRange(this,8,999)"/>&nbsp;&nbsp;
-						<input type="checkbox" id="show_pass1" onclick="togglePass('wifi_pass1')"/>
-						<label for="show_pass1" id="show_pass1_label"><%~ rvel %></label><br/>
+						<input type="text" id="wifi_pass1" class="form-control" size="20" onkeyup="proofreadLengthRange(this,8,999)"/>&nbsp;&nbsp;
 					</span>
 				</div>
 
@@ -1044,9 +1062,7 @@ var isb43 = wirelessDriver == "mac80211" && (!GwifiN) ? true : false ;
 					<div id="wifi_guest_pass1_container" class="row indent">
 						<label class="col-xs-5" for="wifi_guest_pass1" id="wifi_guest_pass1_label"><%~ Pswd %>:</label>
 						<span class="col-xs-7">
-							<input type="password" id="wifi_guest_pass1" class="form-control" size="20" onkeyup="proofreadLengthRange(this,8,999)"/>&nbsp;&nbsp;
-							<input type="checkbox" id="show_guest_pass1" onclick="togglePass('wifi_guest_pass1')"/>
-							<label for="show_guest_pass1" id="show_guest_pass1_label"><%~ rvel %></label><br/>
+							<input type="text" id="wifi_guest_pass1" class="form-control" size="20" onkeyup="proofreadLengthRange(this,8,999)"/>&nbsp;&nbsp;
 						</span>
 					</div>
 
