--- a/configure.ac	2015-09-24 19:16:47.233024664 -0400
+++ b/configure.ac	2015-09-24 19:18:48.167063883 -0400
@@ -326,6 +326,7 @@
 compile_crypto=false
 if test "$enable_crypto" != "no"; then
 	have_libgcrypt=false
+	m4_ifdef([AM_PATH_LIBGCRYPT],
 	AM_PATH_LIBGCRYPT(1.2.2, [ have_libgcrypt=true ],
 	[
 		if test "$enable_crypto" = "yes"; then
@@ -333,7 +334,7 @@
 		else
 			AC_MSG_WARN([ntfsprogs crypto code requires the gcrypt library.])
 		fi
-	])
+	]),[have_libgcrypt=false])
 	have_libgnutls=false
 	PKG_CHECK_MODULES(GNUTLS, gnutls >= 1.4.4, [ have_libgnutls=true ],
 		if test "$enable_crypto" = "yes"; then
