--- a/easy-rsa/2.0/openssl-1.0.0.cnf	2014-12-10 15:30:33.000000000 -0500
+++ b/easy-rsa/2.0/openssl-1.0.0.cnf	2014-12-13 10:58:22.541543957 -0500
@@ -68,7 +68,7 @@
 stateOrProvinceName	= match
 organizationName	= match
 organizationalUnitName	= optional
-commonName		= supplied
+commonName		= optional
 name			= optional
 emailAddress		= optional
 
@@ -81,7 +81,7 @@
 localityName		= optional
 organizationName	= optional
 organizationalUnitName	= optional
-commonName		= supplied
+commonName		= optional
 name			= optional
 emailAddress		= optional
 
