--- a/easy-rsa/2.0/openssl-1.0.0.cnf
+++ b/easy-rsa/2.0/openssl-1.0.0.cnf
@@ -1,9 +1,8 @@
-# For use with easy-rsa version 2.0 and OpenSSL 1.0.0*
+# For use with easy-rsa version 2.0 and OpenSSL 1.1.*
 
 # This definition stops the following lines choking if HOME isn't
 # defined.
 HOME			= .
-RANDFILE		= $ENV::HOME/.rnd
 openssl_conf		= openssl_init
 
 [ openssl_init ]
--- a/easy-rsa/2.0/whichopensslcnf
+++ b/easy-rsa/2.0/whichopensslcnf
@@ -9,6 +9,8 @@
 		cnf="$1/openssl-0.9.8.cnf"
 	elif $OPENSSL version | grep -E "1\.0\.[[:digit:]][[:alnum:]]?" > /dev/null; then
                 cnf="$1/openssl-1.0.0.cnf"
+	elif $OPENSSL version | grep -E "1\.1\.[[:digit:]][[:alnum:]]?" > /dev/null; then
+                cnf="$1/openssl-1.1.0.cnf"
 	else
 		cnf="$1/openssl.cnf"
 	fi
