--- kamikaze-8.09-src/include/image.mk	2009-01-04 15:35:03.000000000 -0500
+++ kamikaze-8.09-gargoyle-src/include/image.mk	2009-01-04 18:08:43.000000000 -0500
@@ -108,6 +108,33 @@
 	find $(TARGET_DIR) -type d | $(XARGS) chmod 0755
 	$(INSTALL_DIR) $(TARGET_DIR)/tmp
 	chmod 0777 $(TARGET_DIR)/tmp
+
+
+	#############################################################################
+	############# CUSTOM GARGOYLE CONFIGURATION #################################
+	#############################################################################
+	# we put custom configuration here because this is the absolute last thing called before we build an image
+	# right now it isn't as important that all configuration be done this late, but putting it here makes sure
+	# no other scripts can be called before images are built, modifying or reversing our changes.
+
+	# remove init scripts of things that should not be active by default
+	
+	IPKG_INSTROOT=$(TARGET_DIR) sh $(TARGET_DIR)/etc/rc.common $(TARGET_DIR)/etc/init.d/httpd disable
+	IPKG_INSTROOT=$(TARGET_DIR) sh $(TARGET_DIR)/etc/rc.common $(TARGET_DIR)/etc/init.d/miniupnpd disable
+	IPKG_INSTROOT=$(TARGET_DIR) sh $(TARGET_DIR)/etc/rc.common $(TARGET_DIR)/etc/init.d/qos_gargoyle disable
+	IPKG_INSTROOT=$(TARGET_DIR) sh $(TARGET_DIR)/etc/rc.common $(TARGET_DIR)/etc/init.d/webmon_gargoyle disable
+	IPKG_INSTROOT=$(TARGET_DIR) sh $(TARGET_DIR)/etc/rc.common $(TARGET_DIR)/etc/init.d/restricter_gargoyle disable
+
+	# ensure ntpclient is active
+	chmod 755 $(TARGET_DIR)/etc/hotplug.d/iface/20-ntpclient
+
+	# by default, update time every 12 hours
+	mkdir -p $(TARGET_DIR)/etc/crontabs
+	echo '0	0,12	*	*	*	ACTION=ifup /etc/hotplug.d/iface/20-ntpclient' >> $(TARGET_DIR)/etc/crontabs/root
+
+	##############################################################################
+
+
 endef
 
 define Image/mkfs/prepare
