Index: Makefile.openwrt
===================================================================
RCS file: /usr/local/cvsroot/miniupnp/miniupnpd/Makefile.openwrt,v
retrieving revision 1.11
diff -u -r1.11 miniupnpd-1.1/Makefile.openwrt
--- miniupnpd-1.1.orig/Makefile.openwrt	29 Jan 2008 12:52:03 -0000	1.11
+++ miniupnpd-1.1/Makefile.openwrt	1 May 2008 22:34:28 -0000
@@ -6,7 +6,7 @@
 # your openwrt SDK/etc...
 # http://miniupnp.tuxfamily.org/forum/viewtopic.php?p=713
 
-include $(TOPDIR)/rules.mk
+#include $(TOPDIR)/rules.mk
 #CFLAGS = -Wall -O -D_GNU_SOURCE -g
 #CFLAGS = -Wall -Os -D_GNU_SOURCE
 #CC = mipsel-linux-gcc
@@ -16,7 +16,8 @@
 CP = cp
 RM = rm -f
 INSTALL = install
-IPTABLESINC=$(BUILD_DIR)/iptables-1.3.3/include
+IPTABLESINC=$(STAGING_DIR)/usr/include
+#IPTABLESINC=$(BUILD_DIR)/iptables-1.3.3/include
 #IPTABLESINC=$(TOPDIR)/../iptables-1.3.3/include
 
 BASEOBJS = miniupnpd.o upnphttp.o upnpdescgen.o upnpsoap.o \
@@ -28,7 +29,7 @@
 NETFILTEROBJS = netfilter/iptcrdr.o
 ALLOBJS = $(BASEOBJS) $(LNXOBJS) $(NETFILTEROBJS)
 
-LIBS = $(BUILD_DIR)/iptables-1.3.3/libiptc/libiptc.a
+#LIBS = $(BUILD_DIR)/iptables-1.3.3/libiptc/libiptc.a
 #LIBS = $(TOPDIR)/../iptables-1.3.3/libiptc/libiptc.a
 
 TESTUPNPDESCGENOBJS = testupnpdescgen.o upnpdescgen.o
