Index: miniupnpd-1.1/netfilter/iptables_init.sh
===================================================================
--- miniupnpd-1.1.orig/netfilter/iptables_init.sh	2007-09-17 17:43:54.000000000 -0500
+++ miniupnpd-1.1/netfilter/iptables_init.sh	2007-09-17 17:43:54.000000000 -0500
@@ -1,9 +1,12 @@
 #! /bin/sh
 # $Id: iptables_init.sh,v 1.3 2006/11/23 12:32:57 nanard Exp $
+. /etc/functions.sh
 IPTABLES=iptables
 
 #change this parameters :
-EXTIF=eth0
+include /lib/network
+scan_interfaces
+config_get EXTIF wan ifname
 EXTIP="`LC_ALL=C /sbin/ifconfig $EXTIF | grep 'inet addr' | awk '{print $2}' | sed -e 's/.*://'`"
 echo "External IP = $EXTIP"
 
Index: miniupnpd-1.1/netfilter/iptables_removeall.sh
===================================================================
--- miniupnpd-1.1.orig/netfilter/iptables_removeall.sh	2007-09-17 17:43:54.000000000 -0500
+++ miniupnpd-1.1/netfilter/iptables_removeall.sh	2007-09-17 17:43:54.000000000 -0500
@@ -1,9 +1,12 @@
 #! /bin/sh
 # $Id: iptables_removeall.sh,v 1.3 2006/11/23 12:32:57 nanard Exp $
+. /etc/functions.sh
 IPTABLES=iptables
 
 #change this parameters :
-EXTIF=eth0
+include /lib/network
+scan_interfaces
+config_get EXTIF wan ifname
 EXTIP="`LC_ALL=C /sbin/ifconfig $EXTIF | grep 'inet addr' | awk '{print $2}' | sed -e 's/.*://'`"
 
 #removing the MINIUPNPD chain for nat
