Index: miniupnpd-1.1/config.h.openwrt
===================================================================
RCS file: /usr/local/cvsroot/miniupnp/miniupnpd/miniupnpd-1.1/config.h.openwrt,v
retrieving revision 1.2
diff -u -r1.2 miniupnpd-1.1/config.h.openwrt
--- miniupnpd-1.1/config.h.openwrt	26 Apr 2007 19:11:09 -0000	1.2
+++ miniupnpd-1.1/config.h.openwrt	1 May 2008 22:47:17 -0000
@@ -1,18 +1,27 @@
 /* $Id: miniupnpd-1.1/config.h.openwrt,v 1.2 2007/04/26 19:11:09 nanard Exp $ */
 /* MiniUPnP Project
  * http:///miniupnp.free.fr
- * (c) 2006 Thomas Bernard
+ * (c) 2006-2008 Thomas Bernard
  */
 #ifndef __CONFIG_H__
 #define __CONFIG_H__
 
 #define UPNP_VERSION	"20070228"
-
+#define USE_NETFILTER 1
 #define OS_NAME	"OpenWRT"
 /* OS/version */
-#define OS_VERSION	"OpenWRT/WhiteRussian"
+#define OS_VERSION	"OpenWRT/kamikaze"
 #define OS_URL	"http://openwrt.org/"
 
 #define LOG_MINIUPNPD	LOG_DAEMON
+
+/* Comment the following line to disable NAT-PMP operations */
+#define ENABLE_NATPMP
+
+/* Comment the following line to use home made daemonize() func instead
+ * of BSD daemon() */
+#define USE_DAEMON
+
+#define ENABLE_L3F_SERVICE
 
 #endif
