--- a/getifstats.h	2012-10-04 18:34:08.000000000 -0400
+++ b/getifstats.h	2015-06-17 15:42:38.599113280 -0400
@@ -8,11 +8,11 @@
 #define GETIFSTATS_H_INCLUDED
 
 struct ifdata {
-	unsigned long opackets;
-	unsigned long ipackets;
-	unsigned long obytes;
-	unsigned long ibytes;
-	unsigned long baudrate;
+	uint64_t opackets;
+	uint64_t ipackets;
+	uint64_t obytes;
+	uint64_t ibytes;
+	uint64_t baudrate;
 };
 
 /* getifstats()
--- a/upnpsoap.c	2015-03-07 10:56:51.000000000 -0500
+++ b/upnpsoap.c	2015-06-17 16:02:44.999999593 -0400
@@ -18,6 +18,8 @@
 #include <arpa/inet.h>
 #include <netdb.h>
 
+
+
 #include "macros.h"
 #include "config.h"
 #include "upnpglobalvars.h"
@@ -31,6 +33,13 @@
 #include "getconnstatus.h"
 #include "upnpurns.h"
 
+
+#ifndef PRIu64
+#define PRIu64 "llu"
+#endif
+
+
+
 static void
 BuildSendAndCloseSoapResp(struct upnphttp * h,
                           const char * body, int bodylen)
@@ -86,7 +95,7 @@
 	static const char resp[] =
 		"<u:%sResponse "
 		"xmlns:u=\"%s\">"
-		"<NewTotalBytesSent>%lu</NewTotalBytesSent>"
+		"<NewTotalBytesSent>%" PRIu64 "</NewTotalBytesSent>"
 		"</u:%sResponse>";
 
 	char body[512];
@@ -108,7 +117,7 @@
 	static const char resp[] =
 		"<u:%sResponse "
 		"xmlns:u=\"%s\">"
-		"<NewTotalBytesReceived>%lu</NewTotalBytesReceived>"
+		"<NewTotalBytesReceived>%" PRIu64 "</NewTotalBytesReceived>"
 		"</u:%sResponse>";
 
 	char body[512];
@@ -130,7 +139,7 @@
 	static const char resp[] =
 		"<u:%sResponse "
 		"xmlns:u=\"%s\">"
-		"<NewTotalPacketsSent>%lu</NewTotalPacketsSent>"
+		"<NewTotalPacketsSent>%" PRIu64 "</NewTotalPacketsSent>"
 		"</u:%sResponse>";
 
 	char body[512];
@@ -152,7 +161,7 @@
 	static const char resp[] =
 		"<u:%sResponse "
 		"xmlns:u=\"%s\">"
-		"<NewTotalPacketsReceived>%lu</NewTotalPacketsReceived>"
+		"<NewTotalPacketsReceived>%" PRIu64 "</NewTotalPacketsReceived>"
 		"</u:%sResponse>";
 
 	char body[512];
@@ -175,8 +184,8 @@
 		"<u:%sResponse "
 		"xmlns:u=\"%s\">"
 		"<NewWANAccessType>%s</NewWANAccessType>"
-		"<NewLayer1UpstreamMaxBitRate>%lu</NewLayer1UpstreamMaxBitRate>"
-		"<NewLayer1DownstreamMaxBitRate>%lu</NewLayer1DownstreamMaxBitRate>"
+		"<NewLayer1UpstreamMaxBitRate>%" PRIu64 "</NewLayer1UpstreamMaxBitRate>"
+		"<NewLayer1DownstreamMaxBitRate>%" PRIu64 "</NewLayer1DownstreamMaxBitRate>"
 		"<NewPhysicalLinkStatus>%s</NewPhysicalLinkStatus>"
 		"</u:%sResponse>";
 
--- a/linux/getifstats.c	2015-06-09 11:36:22.000000000 -0400
+++ b/linux/getifstats.c	2015-06-17 17:47:27.472425710 -0400
@@ -24,6 +24,38 @@
 /* that is the answer */
 #define BAUDRATE_DEFAULT 4200000
 
+/* from <inttypes.h> */
+#ifndef PRIu64
+#define PRIu64 "llu"
+#endif
+
+
+
+uint64_t strtou64(const char *str, char** end )
+{
+	uint64_t val = 0;
+	int found = sscanf(str, "%" PRIu64, &val);
+	if(found == 0)
+	{
+		val = 0;
+		*end = (char*)str;
+	}
+	else
+	{
+		char* endp = (char*)str;
+		while( endp[0] != '\0' && (endp[0] < '0' || endp[0] > '9') )
+		{
+			endp++;
+		}
+		while( endp[0] != '\0' && endp[0] >= '0' && endp[0] <= '9')
+		{
+			endp++;
+		}
+		*end = endp;
+	}
+	return val;
+}
+
 int
 getifstats(const char * ifname, struct ifdata * data)
 {
@@ -80,18 +112,18 @@
 			continue;
 		p++;
 		while(*p==' ') p++;
-		data->ibytes = strtoul(p, &p, 0);
+		data->ibytes = strtou64(p, &p);
 		while(*p==' ') p++;
-		data->ipackets = strtoul(p, &p, 0);
+		data->ipackets = strtou64(p, &p);
 		/* skip 6 columns */
 		for(i=6; i>0 && *p!='\0'; i--) {
 			while(*p==' ') p++;
 			while(*p!=' ' && *p) p++;
 		}
 		while(*p==' ') p++;
-		data->obytes = strtoul(p, &p, 0);
+		data->obytes = strtou64(p, &p);
 		while(*p==' ') p++;
-		data->opackets = strtoul(p, &p, 0);
+		data->opackets = strtou64(p, &p);
 		r = 0;
 		break;
 	}
@@ -117,7 +149,7 @@
 		if(s >= 0) {
 			strncpy(iwr.ifr_name, ifname, IFNAMSIZ);
 			if(ioctl(s, SIOCGIWRATE, &iwr) >= 0) {
-				data->baudrate = iwr.u.bitrate.value;
+				data->baudrate = (uint64_t)(iwr.u.bitrate.value);
 			}
 			close(s);
 		}
