--- kamikaze-7.09-src/include/verbose.mk	2007-09-20 04:53:16.000000000 -0400
+++ kamikaze-7.09-gargoyle-src/include/verbose.mk	2008-11-29 22:01:38.000000000 -0500
@@ -38,7 +38,7 @@
   else
     export QUIET:=1
     ifeq ($(KBUILD_VERBOSE),0)
-      MAKE:=&>/dev/null $(MAKE)
+      MAKE:= >/dev/null 2>&1 $(MAKE)
     endif
     MAKE:=cmd() { $(MAKE) $$* || {  echo "Build failed. Please re-run make with V=99 to see what's going on"; false; } } 3>&1 4>&2; cmd
   endif
