--- ../package/gargoyle/Makefile	2008-09-01 15:35:54.000000000 -0400
+++ package/gargoyle/Makefile	2008-09-01 15:38:40.000000000 -0400
@@ -14,7 +14,7 @@
 	SECTION:=admin
 	CATEGORY:=Administration
 	TITLE:=Gargoyle Web Interface
-	DEPENDS:=+uci +libuci +haserl-dev +ntpclient +miniupnpd +httpd-gargoyle +qos-gargoyle +bwmon-gargoyle +ddns-gargoyle +libericstools 
+	DEPENDS:=+wpa-supplicant +hostapd +uci +libuci +haserl-dev +ntpclient +miniupnpd +httpd-gargoyle +qos-gargoyle +bwmon-gargoyle +ddns-gargoyle +libericstools 
 	MAINTAINER:=Eric Bishop <eric@gargoyle-router.com>
 endef
 
