--- kamikaze-8.09-src/package/base-files/files/etc/passwd	2009-01-04 15:18:03.000000000 -0500
+++ kamikaze-8.09-gargoyle-src/package/base-files/files/etc/passwd	2009-01-04 15:24:42.000000000 -0500
@@ -1,3 +1,3 @@
-root:!:0:0:root:/root:/bin/ash
+root:$1$V4UetPzk$CYXluq4wUazHjmCDBCqXF.:0:0:root:/root:/bin/ash
 nobody:*:65534:65534:nobody:/var:/bin/false
 daemon:*:65534:65534:daemon:/var:/bin/false
