--- backfire-src/include/image.mk	2010-04-20 11:13:42.000000000 -0400
+++ atheros-2.6-src/include/image.mk	2010-04-22 12:54:10.000000000 -0400
@@ -131,6 +131,29 @@
 	- $(FIND) $(TARGET_DIR) -type d -print0 | $(XARGS) -0 chmod 0755
 	$(INSTALL_DIR) $(TARGET_DIR)/tmp
 	chmod 0777 $(TARGET_DIR)/tmp
+
+
+	#############################################################################
+	############# CUSTOM GARGOYLE CONFIGURATION #################################
+	#############################################################################
+	# we put custom configuration here because this is the absolute last thing called before we build an image
+	# right now it isn't as important that all configuration be done this late, but putting it here makes sure
+	# no other scripts can be called before images are built, modifying or reversing our changes.
+
+	# remove init scripts of things that should not be active by default
+	
+	IPKG_INSTROOT=$(TARGET_DIR) sh $(TARGET_DIR)/etc/rc.common $(TARGET_DIR)/etc/init.d/httpd disable
+	IPKG_INSTROOT=$(TARGET_DIR) sh $(TARGET_DIR)/etc/rc.common $(TARGET_DIR)/etc/init.d/miniupnpd disable
+	IPKG_INSTROOT=$(TARGET_DIR) sh $(TARGET_DIR)/etc/rc.common $(TARGET_DIR)/etc/init.d/qos_gargoyle disable
+	IPKG_INSTROOT=$(TARGET_DIR) sh $(TARGET_DIR)/etc/rc.common $(TARGET_DIR)/etc/init.d/webmon_gargoyle disable
+
+	# ensure ntpclient is active
+	chmod 755 $(TARGET_DIR)/etc/hotplug.d/iface/20-ntpclient
+
+
+	##############################################################################
+
+
 endef
 
 define Image/mkfs/prepare
