--- a/package/base-files/files/etc/sysctl.conf	2012-10-20 10:59:16.750492941 -0400
+++ b/package/base-files/files/etc/sysctl.conf	2012-10-20 11:03:30.290833414 -0400
@@ -16,8 +16,8 @@
 
 net.netfilter.nf_conntrack_acct=1
 net.netfilter.nf_conntrack_checksum=0
-net.netfilter.nf_conntrack_max=16384
-net.netfilter.nf_conntrack_tcp_timeout_established=3600
+#net.netfilter.nf_conntrack_max=16384
+#net.netfilter.nf_conntrack_tcp_timeout_established=3600
 net.netfilter.nf_conntrack_udp_timeout=60
 net.netfilter.nf_conntrack_udp_timeout_stream=180
 
@@ -25,3 +25,8 @@
 net.bridge.bridge-nf-call-arptables=0
 net.bridge.bridge-nf-call-ip6tables=0
 net.bridge.bridge-nf-call-iptables=0
+
+net.ipv4.tcp_tw_reuse=1
+net.ipv4.tcp_max_tw_buckets=8192
+net.netfilter.nf_conntrack_max=4096
+net.netfilter.nf_conntrack_tcp_timeout_established=600
