--- a/target/linux/ramips/base-files/lib/preinit/07_set_preinit_iface_ramips
+++ b/target/linux/ramips/base-files/lib/preinit/07_set_preinit_iface_ramips
@@ -26,6 +26,11 @@
 		vconfig add eth0 1
 		ifconfig eth0 up
 		ifname=eth0.1
+	elif grep MT7620 /proc/cpuinfo; then
+		swconfig dev mt7620 set reset 1
+		swconfig dev mt7620 set enable_vlan 0
+		swconfig dev mt7620 set apply
+		ifname=eth0
 	else
 		ifname=eth0
 	fi
