--- a/src/crypt.c
+++ b/src/crypt.c
@@ -167,9 +167,9 @@ crypt_set_iv(crypt_t *key, const uchar *
 void
 stream_crypt(crypt_t *key, uchar *buf, size_t len)
 {
-  AES_ctr128_encrypt(buf, buf, len,
+  CRYPTO_ctr128_encrypt(buf, buf, len,
                      &key->key, key->ivec, key->ecount_buf,
-                     &key->pos);
+                     &key->pos, (block128_f)AES_encrypt);
 }
 
 /**
--- a/src/crypt.h
+++ b/src/crypt.h
@@ -56,6 +56,7 @@ int random_int(unsigned int max);
 #ifdef CRYPT_PRIVATE
 
 #include <openssl/aes.h>
+#include <openssl/modes.h>
 
 /* ==========
    These definitions are not part of the crypt interface.
