--- cyassl-1.0.6-orig/src/ssl.c	2009-08-21 12:55:34.000000000 -0400
+++ cyassl-1.0.6/src/ssl.c	2009-08-21 12:56:04.000000000 -0400
@@ -1067,7 +1067,7 @@
 {
     SSL_SESSION* ret = 0;
     const byte*  id = ssl->arrays.sessionID;
-    word32       row, idx;
+    int       row, idx;
 
     if (ssl->options.sessionCacheOff)
         return 0;
--- cyassl-1.0.6-orig/src/cyassl_int.c	2009-08-21 12:55:34.000000000 -0400
+++ cyassl-1.0.6/src/cyassl_int.c	2009-08-21 12:57:34.000000000 -0400
@@ -1427,7 +1427,7 @@
     
     /* Put buffer data at start if not there */
     if (usedLength > 0 && ssl->buffers.inputBuffer.idx != 0)
-        memcpy(ssl->buffers.inputBuffer.buffer,
+        memmove(ssl->buffers.inputBuffer.buffer,
                ssl->buffers.inputBuffer.buffer + ssl->buffers.inputBuffer.idx,
                usedLength);
     
