--- kamikaze-8.09-old/package/ppp/files/ppp.sh	2009-09-11 20:33:50.000000000 -0400
+++ kamikaze-8.09-src/package/ppp/files/ppp.sh	2009-09-11 20:32:26.000000000 -0400
@@ -46,7 +46,16 @@
 	[ "$interval" != "$keepalive" ] || interval=5
 
 	config_get_bool peerdns "$cfg" peerdns 1 
-	[ "$peerdns" -eq 1 ] && peerdns="usepeerdns" || peerdns="" 
+	config_get dns "$cfg" dns
+	if [ -n "$dns" ] ; then 
+		peerdns=0
+	fi
+	
+	if [ "$peerdns" -eq 1 ] ; then
+		peerdns="usepeerdns"
+	 else
+	 	peerdns="" 
+	 fi
 	
 	config_get demand "$cfg" demand
 	[ -n "$demand" ] && echo "nameserver 1.1.1.1" > /tmp/resolv.conf.auto
