--- a/src/or/routerlist.c	2011-12-15 11:27:07.000000000 -0500
+++ b/src/or/routerlist.c	2011-12-19 16:08:22.947282142 -0500
@@ -3679,7 +3679,9 @@
    * networkstatuses, then we should get more before we decide to kill
    * routers.) */
   /* we set this to true iff we don't care about v2 info, or we have enough. */
+  const or_options_t *options = get_options();
   have_enough_v2 = !caches ||
+    !(authdir_mode_any_main(options) || options->FetchV2Networkstatus) ||
     (networkstatus_v2_list &&
      smartlist_len(networkstatus_v2_list) > get_n_v2_authorities() / 2);
 
