--- kamikaze-8.09-src-orig/target/linux/atheros/files/drivers/net/ar2313/ar2313.c	2009-04-06 22:21:38.000000000 -0400
+++ kamikaze-8.09-src/target/linux/atheros/files/drivers/net/ar2313/ar2313.c	2009-04-06 22:24:14.000000000 -0400
@@ -922,7 +922,7 @@
 		printk("RX descr  %08x\n", rxdesc->descr);
 #endif
 
-		if ((status & (DMA_RX_ERROR | DMA_RX_ERR_LENGTH)) &&
+		if ((status & DMA_RX_ERROR) &&
 			(!(status & DMA_RX_LONG))) {
 #if DEBUG_RX
 			printk("%s: rx ERROR %08x\n", __FUNCTION__, status);
