--- a/xattr.c
+++ b/xattr.c
@@ -20,6 +20,16 @@
 #define SYSTEM_NTFS_SECURITY "system.ntfs_security"
 // clang-format on
 
+static inline size_t __must_check size_add(size_t addend1, size_t addend2)
+{
+       size_t bytes;
+
+       if (check_add_overflow(addend1, addend2, &bytes))
+               return SIZE_MAX;
+
+       return bytes;
+}
+
 static inline size_t unpacked_ea_size(const struct EA_FULL *ea)
 {
        return ea->size ? le32_to_cpu(ea->size)

