--- a/package/utils/busybox/Config-defaults.in
+++ b/package/utils/busybox/Config-defaults.in
@@ -462,10 +462,10 @@ config BUSYBOX_DEFAULT_FEATURE_TAR_SELIN
 	default n
 config BUSYBOX_DEFAULT_UNZIP
 	bool
-	default n
+	default y
 config BUSYBOX_DEFAULT_FEATURE_UNZIP_CDF
 	bool
-	default n
+	default y
 config BUSYBOX_DEFAULT_FEATURE_UNZIP_BZIP2
 	bool
 	default n
@@ -669,7 +669,7 @@ config BUSYBOX_DEFAULT_MD5SUM
 	default y
 config BUSYBOX_DEFAULT_SHA1SUM
 	bool
-	default n
+	default y
 config BUSYBOX_DEFAULT_SHA256SUM
 	bool
 	default y
@@ -2176,7 +2176,7 @@ config BUSYBOX_DEFAULT_ARP
 	default n
 config BUSYBOX_DEFAULT_ARPING
 	bool
-	default n
+	default y
 config BUSYBOX_DEFAULT_BRCTL
 	bool
 	default y
