--- a/package/base-files/files/lib/functions.sh
+++ b/package/base-files/files/lib/functions.sh
@@ -154,14 +154,17 @@ default_prerm() {
 	local root="${IPKG_INSTROOT}"
 	local pkgname="$(basename ${1%.*})"
 	local ret=0
+	local pkg_prerm pkg_list
 
-	if [ -f "$root/usr/lib/opkg/info/${pkgname}.prerm-pkg" ]; then
-		( . "$root/usr/lib/opkg/info/${pkgname}.prerm-pkg" )
+	pkg_prerm="${1}-pkg"
+	pkg_list=$(echo "${1}" | sed 's/prerm$/list/g')
+	if [ -f "$pkg_prerm" ]; then
+		( . "$pkg_prerm" )
 		ret=$?
 	fi
 
 	local shell="$(which bash)"
-	for i in $(grep -s "^/etc/init.d/" "$root/usr/lib/opkg/info/${pkgname}.list"); do
+	for i in $(grep -s "/etc/init.d/" "${pkg_list}"); do
 		if [ -n "$root" ]; then
 			${shell:-/bin/sh} "$root/etc/rc.common" "$root$i" disable
 		else
@@ -176,8 +179,8 @@ default_prerm() {
 }
 
 add_group_and_user() {
-	local pkgname="$1"
-	local rusers="$(sed -ne 's/^Require-User: *//p' $root/usr/lib/opkg/info/${pkgname}.control 2>/dev/null)"
+	local pkg_control="$1"
+	local rusers="$(sed -ne 's/^Require-User: *//p' "${pkg_control}" 2>/dev/null)"
 
 	if [ -n "$rusers" ]; then
 		local tuple oIFS="$IFS"
@@ -215,10 +218,19 @@ default_postinst() {
 	local pkgname="$(basename ${1%.*})"
 	local ret=0
 
-	add_group_and_user "${pkgname}"
+	local pkg_postinst="${1}"-pkg
+	local pkg_control=$(echo "${1}" | sed 's/postinst$/control/g')
+	local pkg_list=$(echo "${1}" | sed 's/postinst$/list/g')
+	if [ -n "${root}" ] ; then
+		pkg_postinst="${root}/usr/lib/opkg/info/${pkgname}.postinst-pkg"
+		pkg_control="${root}/usr/lib/opkg/info/${pkgname}.control"
+		pkg_list="${root}/usr/lib/opkg/info/${pkgname}.list"
+	fi
+
+	add_group_and_user "${pkg_control}"
 
-	if [ -f "$root/usr/lib/opkg/info/${pkgname}.postinst-pkg" ]; then
-		( . "$root/usr/lib/opkg/info/${pkgname}.postinst-pkg" )
+	if [ -f "$pkg_postinst" ]; then
+		( . "$pkg_postinst" )
 		ret=$?
 	fi
 
@@ -227,14 +239,14 @@ default_postinst() {
 		rm -fR $root/rootfs-overlay/
 	fi
 
-	if [ -z "$root" ] && grep -q -s "^/etc/modules.d/" "/usr/lib/opkg/info/${pkgname}.list"; then
+	if [ -z "$root" ] && grep -q -s "/etc/modules.d/" "${pkg_list}"; then
 		kmodloader
 	fi
 
-	if [ -z "$root" ] && grep -q -s "^/etc/uci-defaults/" "/usr/lib/opkg/info/${pkgname}.list"; then
+	if [ -z "$root" ] && grep -q -s "/etc/uci-defaults/" "${pkg_list}"; then
 		. /lib/functions/system.sh
 		[ -d /tmp/.uci ] || mkdir -p /tmp/.uci
-		for i in $(grep -s "^/etc/uci-defaults/" "/usr/lib/opkg/info/${pkgname}.list"); do
+		for i in $(grep -s "/etc/uci-defaults/" "${pkg_list}"); do
 			( [ -f "$i" ] && cd "$(dirname $i)" && . "$i" ) && rm -f "$i"
 		done
 		uci commit
@@ -243,7 +255,7 @@ default_postinst() {
 	[ -n "$root" ] || rm -f /tmp/luci-indexcache 2>/dev/null
 
 	local shell="$(which bash)"
-	for i in $(grep -s "^/etc/init.d/" "$root/usr/lib/opkg/info/${pkgname}.list"); do
+	for i in $(grep -s "/etc/init.d/" "${pkg_list}"); do
 		if [ -n "$root" ]; then
 			${shell:-/bin/sh} "$root/etc/rc.common" "$root$i" enable
 		else
