--- a/package/network/config/netifd/files/lib/netifd/dhcp.script
+++ b/package/network/config/netifd/files/lib/netifd/dhcp.script
@@ -4,6 +4,12 @@
 . /lib/functions.sh
 . /lib/netifd/netifd-proto.sh
 
+change_state () {
+	[ -n "$INTERFACE" ] || return
+	uci_revert_state "$1" "$2" "$3" "$4"
+	uci_set_state "$1" "$2" "$3" "$4"
+}
+
 set_classless_routes() {
 	local max=128
 	while [ -n "$1" -a -n "$2" -a $max -gt 0 ]; do
@@ -59,6 +65,31 @@
 	proto_send_update "$INTERFACE"
 
 
+
+	local old_ip
+	local old_broadcast
+	local old_subnet
+	[ -n "$INTERFACE" ] && {
+		old_ip="$(uci_get_state network "$INTERFACE" ipaddr)"
+		old_broadcast="$(uci_get_state network "$INTERFACE" broadcast)"
+		old_subnet="$(uci_get_state network "$INTERFACE" netmask)"
+	}
+	[ "$ip" != "$old_ip" ] \
+	|| [ "${broadcast:-+}" != "$old_broadcast" ] \
+	|| [ "${subnet:-255.255.255.0}" != "$old_subnet" ] && {
+		change_state network "$INTERFACE" ipaddr "$ip"
+		change_state network "$INTERFACE" broadcast "${broadcast:-+}"
+		change_state network "$INTERFACE" netmask "${subnet:-255.255.255.0}"
+	}
+	change_state network "$INTERFACE" gateway "$router"
+	[ -n "$hostname" ] &&	change_state network "$INTERFACE" lease_hostname "$hostname"
+	
+	change_state network "$INTERFACE" lease_server "$serverid" 
+	change_state network "$INTERFACE" lease_lifetime "$lease" 
+	change_state network "$INTERFACE" lease_acquired "$(sed -ne 's![^0-9].*$!!p' /proc/uptime)"
+
+
+
 	if [ "$IFACE6RD" != 0 -a -n "$ip6rd" ]; then
 		local v4mask="${ip6rd%% *}"
 		ip6rd="${ip6rd#* }"
--- /dev/null
+++ b/package/network/services/ppp/files/etc/ppp/ip-down.d/state
@@ -0,0 +1,15 @@
+#!/bin/sh
+. /lib/functions.sh
+PPP_IFACE="$1"
+PPP_TTY="$2"
+PPP_SPEED="$3"
+PPP_LOCAL="$4"
+PPP_REMOTE="$5"
+PPP_IPPARAM="$(echo $6 | sed 's/\./_/g')"
+export PPP_IFACE PPP_TTY PPP_SPEED PPP_LOCAL PPP_REMOTE PPP_IPPARAM
+[ -z "$PPP_IPPARAM" ] || {
+
+	# remove the interface's network state
+	uci_revert_state network "$PPP_IPPARAM"
+}
+
--- /dev/null
+++ b/package/network/services/ppp/files/etc/ppp/ip-up.d/state
@@ -0,0 +1,22 @@
+#!/bin/sh
+
+. /lib/functions.sh
+. /lib/network/config.sh
+
+PPP_IFACE="$1"
+PPP_TTY="$2"
+PPP_SPEED="$3"
+PPP_LOCAL="$4"
+PPP_REMOTE="$5"
+PPP_IPPARAM="$(echo $6 | sed 's/\./_/g')"
+
+export PPP_IFACE PPP_TTY PPP_SPEED PPP_LOCAL PPP_REMOTE PPP_IPPARAM
+[ -z "$PPP_IPPARAM" -o -z "$PPP_LOCAL" ] || {
+	uci_set_state network "$PPP_IPPARAM" ifname "$PPP_IFACE"
+	uci_set_state network "$PPP_IPPARAM" ipaddr "$PPP_LOCAL"
+	uci_set_state network "$PPP_IPPARAM" gateway "$PPP_REMOTE"
+	
+	# Point to Point link, Netmask should be 255.255.255.255
+	uci_set_state network "$PPP_IPPARAM" netmask "255.255.255.255"
+}
+
--- a/package/network/services/ppp/Makefile
+++ b/package/network/services/ppp/Makefile
@@ -223,9 +223,13 @@
 	$(INSTALL_DIR) $(1)/usr/sbin
 	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/sbin/pppd $(1)/usr/sbin/
 	$(INSTALL_DIR) $(1)/etc/ppp
+	$(INSTALL_DIR) $(1)/etc/ppp/ip-up.d
+	$(INSTALL_DIR) $(1)/etc/ppp/ip-down.d
 	$(INSTALL_CONF) ./files/etc/ppp/chap-secrets $(1)/etc/ppp/
 	$(INSTALL_DATA) ./files/etc/ppp/filter $(1)/etc/ppp/
 	$(INSTALL_DATA) ./files/etc/ppp/options $(1)/etc/ppp/
+	$(INSTALL_BIN) ./files/etc/ppp/ip-up.d/state $(1)/etc/ppp/ip-up.d/
+	$(INSTALL_BIN) ./files/etc/ppp/ip-down.d/state $(1)/etc/ppp/ip-down.d/
 	$(LN) /tmp/resolv.conf.ppp $(1)/etc/ppp/resolv.conf
 	$(INSTALL_DIR) $(1)/lib/netifd/proto
 	$(INSTALL_BIN) ./files/ppp.sh $(1)/lib/netifd/proto/

