--- a/package/network/config/netifd/files/lib/netifd/dhcp.script
+++ b/package/network/config/netifd/files/lib/netifd/dhcp.script
@@ -60,6 +60,8 @@
 	[ -n "$message" ]  && json_add_string message "$message"
 	[ -n "$timezone" ] && json_add_int timezone "$timezone"
 	[ -n "$lease" ]    && json_add_int leasetime "$lease"
+	[ -n "$serverid" ] && json_add_string leaseserver "$serverid"
+	json_add_int leaseacquired "$(sed -ne 's![^0-9].*$!!p' /proc/uptime)"
 	proto_close_data
 
 	proto_send_update "$INTERFACE"
