--- a/include/image.mk
+++ b/include/image.mk
@@ -168,8 +168,11 @@ DTC_FLAGS += \
   -Wno-unique_unit_address
 
 define Image/pad-to
-	dd if=$(1) of=$(1).new bs=$(2) conv=sync
-	mv $(1).new $(1)
+	if [ -e "$(1)" ]; then \
+		dd if=$(1) of=$(1).new bs=$(2) conv=sync; \
+		mv $(1).new $(1); \
+	fi; \
+	echo ""
 endef
 
 ROOTFS_PARTSIZE=$(shell echo $$(($(CONFIG_TARGET_ROOTFS_PARTSIZE)*1024*1024)))
--- a/target/linux/ramips/image/Makefile
+++ b/target/linux/ramips/image/Makefile
@@ -84,8 +84,11 @@ define Build/mkdlinkfw-factory
 		-F $@ \
 		-o $@.new \
 		$(if $(DLINK_IMAGE_OFFSET), -O $(DLINK_IMAGE_OFFSET)) \
-		-s $(DLINK_FIRMWARE_SIZE)
-	mv $@.new $@
+		-s $(DLINK_FIRMWARE_SIZE); \
+	if [ -e "$@.new" ]; then \
+		mv $@.new $@; \
+	fi; \
+	echo ""
 endef
 
 define Build/mkdlinkfw-loader
@@ -94,8 +97,11 @@ define Build/mkdlinkfw-loader
 		-r $@ \
 		-o $@.new \
 		$(if $(DLINK_IMAGE_OFFSET), -O $(DLINK_IMAGE_OFFSET)) \
-		-s $(DLINK_FIRMWARE_SIZE)
-	mv $@.new $@
+		-s $(DLINK_FIRMWARE_SIZE); \
+	if [ -e "$@.new" ]; then \
+		mv $@.new $@; \
+	fi; \
+	echo ""
 endef
 
 define Build/netis-tail
