--- 15.05-a03a846/target/linux/ramips/image/Makefile	2016-03-04 12:12:00.906474059 -0500
+++ ramips-src/target/linux/ramips/image/Makefile	2016-03-04 14:32:06.132221069 -0500
@@ -873,10 +873,12 @@
 	cat ex2700-fakeroot.uImage >> $(KDIR)/vmlinux-$(2).uImage
 	$(call MkImageSysupgrade/squashfs,squashfs,$(2),$(6))
 
-	$(STAGING_DIR_HOST)/bin/mkdniimg \
+	if [ -e $(call imgname,squashfs,$(2))-sysupgrade.bin ] && [ -e $(call imgname,squashfs,$(2))-factory.bin ] ; then \
+		$(STAGING_DIR_HOST)/bin/mkdniimg \
 		-B $(3) -H $(5) -v OpenWrt \
 		-i $(call imgname,squashfs,$(2))-sysupgrade.bin \
-		-o $(call imgname,squashfs,$(2))-factory.bin
+		-o $(call imgname,squashfs,$(2))-factory.bin ;\
+	fi
 endef
 
 Image/Build/Profile/E1700=$(call BuildFirmware/UMedia/$(1),$(1),e1700,E1700,0x013326)
--- a/target/linux/ar71xx/image/Makefile	2015-07-21 12:50:17.095881121 -0400
+++ b/target/linux/ar71xx/image/Makefile	2015-09-23 19:40:32.593342825 -0400
@@ -1125,7 +1125,8 @@
 				-o $(call imgname,$(1),$(2))-factory$$$$dashr.bin; \
 			true; \
 		done; \
-	fi
+	fi ; \
+	echo ""
 endef
 
 Image/Build/CameoAP121/buildkernel=$(call MkuImageLzma,$(2),$(3) $(cameo_ap121_mtdlayout))
@@ -1273,7 +1274,8 @@
 	-$(STAGING_DIR_HOST)/bin/mkfwimage2 -f 0x9f000000 \
 		-v $(5).$(6).v6.0.0-OpenWrt-$(REVISION) \
 		-p jffs2:0x50000:0xf60000:0:0:$(call sysupname,$(1),$(2)) \
-		-o $(call factoryname,$(1),$(2))
+		-o $(call factoryname,$(1),$(2)) \
+	echo ""
 endef
 
 # $(4) = board (XS2, XS5, RS, XM)
@@ -1290,7 +1292,8 @@
 			-k $(KDIR_TMP)/$(2)-mtdpart-kernel.bin \
 			-r $(KDIR_TMP)/$(2)-mtdpart-rootfs.bin \
 			-o $(call factoryname,$(1),$(2)); \
-	fi
+	fi ; \
+	echo ""
 endef
 
 Image/Build/UBNTXM/initramfs=$(call MkuImageLzma/initramfs,$(2),$(3) $(ubntxm_mtdlayout))
@@ -1508,7 +1511,8 @@
 	-$(STAGING_DIR_HOST)/bin/mktplinkfw -c \
 		-H $(4) -W $(5) -F $(6) -N OpenWrt -V $(REVISION) $(7) -s \
 		-k $(KDIR_TMP)/vmlinux-initramfs-$(2).bin.lzma \
-		-o $(call imgname,$(1),$(2))-uImage.bin
+		-o $(call imgname,$(1),$(2))-uImage.bin \
+	echo ""
 endef
 
 
@@ -1536,18 +1540,19 @@
 
 define Image/Build/CyberTAN
 	echo -n '' > $(KDIR_TMP)/empty.bin
-	$(STAGING_DIR_HOST)/bin/trx -o $(KDIR)/image.tmp \
+	-$(STAGING_DIR_HOST)/bin/trx -o $(KDIR)/image.tmp \
 		-f $(KDIR_TMP)/vmlinux-$(2).uImage -F $(KDIR_TMP)/empty.bin \
 		-x 32 -a 0x10000 -x -32 -f $(KDIR)/root.$(1)
 	-$(STAGING_DIR_HOST)/bin/addpattern -B $(2) -v v$(5) \
 		-i $(KDIR)/image.tmp \
 		-o $(call sysupname,$(1),$(2))
-	$(STAGING_DIR_HOST)/bin/trx -o $(KDIR)/image.tmp -f $(KDIR_TMP)/vmlinux-$(2).uImage \
+	-$(STAGING_DIR_HOST)/bin/trx -o $(KDIR)/image.tmp -f $(KDIR_TMP)/vmlinux-$(2).uImage \
 		-x 32 -a 0x10000 -x -32 -f $(KDIR)/root.$(1)
 	-$(STAGING_DIR_HOST)/bin/addpattern -B $(2) -v v$(5) -g \
 		-i $(KDIR)/image.tmp \
 		-o $(call factoryname,$(1),$(2))
-	rm $(KDIR)/image.tmp
+	rm -rf $(KDIR)/image.tmp ; \
+	echo ""
 endef
 
 Image/Build/CyberTANGZIP/loader=$(call Image/BuildLoader,$(1),gz,$(2),0x80060000)
@@ -1741,7 +1746,8 @@
 			-k $(KDIR_TMP)/vmlinux-$(2).uImage \
 			-r $(BIN_DIR)/$(IMG_PREFIX)-root.$(1) \
 			-o $(call imgname,$(1),$(2))-factory.img; \
-	fi
+	fi ; \
+	echo ""
 endef
 
 
