--- a/package/utils/busybox/files/sysntpd	2018-03-13 11:01:52.687072947 -0400
+++ b/package/utils/busybox/files/sysntpd	2018-03-13 11:07:23.259916820 -0400
@@ -1,7 +1,7 @@
 #!/bin/sh /etc/rc.common
 # Copyright (C) 2011 OpenWrt.org
 
-START=98
+START=41
 
 USE_PROCD=1
 PROG=/usr/sbin/ntpd
--- a/package/base-files/files/bin/config_generate	2018-03-13 11:01:52.691073023 -0400
+++ b/package/base-files/files/bin/config_generate	2018-03-13 11:06:07.671028273 -0400
@@ -233,11 +233,11 @@
 		delete system.ntp
 		set system.ntp='timeserver'
 		set system.ntp.enabled='1'
-		set system.ntp.enable_server='0'
-		add_list system.ntp.server='0.openwrt.pool.ntp.org'
-		add_list system.ntp.server='1.openwrt.pool.ntp.org'
-		add_list system.ntp.server='2.openwrt.pool.ntp.org'
-		add_list system.ntp.server='3.openwrt.pool.ntp.org'
+		set system.ntp.enable_server='1'
+		add_list system.ntp.server='0.pool.ntp.org'
+		add_list system.ntp.server='1.pool.ntp.org'
+		add_list system.ntp.server='2.pool.ntp.org'
+		add_list system.ntp.server='3.pool.ntp.org'
 	EOF
 
 	if json_is_a system object; then
