--- backfire-src/package/base-files/files/etc/sysctl.conf	2010-04-22 13:08:34.000000000 -0400
+++ atheros-2.6-src/package/base-files/files/etc/sysctl.conf	2010-04-22 15:04:40.000000000 -0400
@@ -12,8 +12,6 @@
 net.core.netdev_max_backlog=30
 net.netfilter.nf_conntrack_checksum=0
 net.ipv4.netfilter.ip_conntrack_checksum=0
-net.ipv4.netfilter.ip_conntrack_max=16384
-net.ipv4.netfilter.ip_conntrack_tcp_timeout_established=3600
 net.ipv4.netfilter.ip_conntrack_udp_timeout=60
 net.ipv4.netfilter.ip_conntrack_udp_timeout_stream=180
 # net.ipv6.conf.all.forwarding=1
@@ -22,3 +20,9 @@
 net.bridge.bridge-nf-call-arptables=0
 net.bridge.bridge-nf-call-ip6tables=0
 net.bridge.bridge-nf-call-iptables=0
+
+net.ipv4.tcp_tw_reuse=1
+net.ipv4.netfilter.ip_conntrack_max=4096
+net.ipv4.netfilter.ip_conntrack_tcp_timeout_established=600
+net.ipv4.netfilter.ip_conntrack_udp_timeout_stream=180
+
