--- a/include/rootfs.mk	2018-03-13 11:01:52.675072718 -0400
+++ b/include/rootfs.mk	2018-03-13 15:10:48.123732989 -0400
@@ -55,6 +55,14 @@
   endef
 endif
 
+ifdef CONFIG_CLEAN_IPKG_PARTIAL
+  define clean_ipkg
+	-find $(1)/usr/lib/opkg/info/* -type f -and -not -name '*.control' -and -not -name 'plugin-gargoyle-*' | $(XARGS) rm -rf
+	-sed -i -ne '/^Require-User: /p' $(1)/usr/lib/opkg/info/*.control
+	-find $(1)/usr/lib/opkg -empty | $(XARGS) rm -rf
+  endef
+endif
+
 define prepare_rootfs
 	@if [ -d $(TOPDIR)/files ]; then \
 		$(call file_copy,$(TOPDIR)/files/.,$(1)); \
--- a/config/Config-build.in
+++ b/config/Config-build.in
@@ -68,6 +68,14 @@ menu "Global build settings"
 		  This removes all ipkg/opkg status data files from the target directory
 		  before building the root filesystem.
 
+	config CLEAN_IPKG_PARTIAL
+		bool
+		prompt "Partially remove ipkg/opkg status data files in final images"
+		default n
+		help
+		  This removes most ipkg/opkg status data files (excluding *.control and plugin-gargoyle-*) from
+		  the target directory before building the root filesystem.
+
 	config INCLUDE_CONFIG
 		bool "Include build configuration in firmware" if DEVEL
 		default n
