--- master-fd30187-bak/target/linux/ar71xx/base-files/lib/ar71xx.sh	2018-03-13 11:01:52.327066070 -0400
+++ master-fd30187/target/linux/ar71xx/base-files/lib/ar71xx.sh	2018-03-13 15:54:49.801191417 -0400
@@ -703,7 +703,7 @@
 		local size="$(mtd_get_part_size 'firmware')"
 
 		[ "$size" = "7929856" ] && name="hornet-ub"
-		[ "$size" = "16318464" ] && name="hornet-ub-x2"
+		[ "$size" = "16318464" ] && name="gargoyle-pocket-router-v2"
 		;;
 	*"JA76PF")
 		name="ja76pf"
--- master-fd30187-bak/target/linux/ar71xx/base-files/lib/upgrade/platform.sh	2018-03-13 11:01:52.327066070 -0400
+++ master-fd30187/target/linux/ar71xx/base-files/lib/upgrade/platform.sh	2018-03-13 15:51:13.756978212 -0400
@@ -248,7 +248,6 @@
 	gl-mifi|\
 	gl-usb150|\
 	hiwifi-hc6361|\
-	hornet-ub-x2|\
 	jwap230|\
 	lima|\
 	loco-m-xw|\
@@ -324,6 +323,7 @@
 	dw33d|\
 	f9k1115v2|\
 	hornet-ub|\
+	gargoyle-pocket-router-v2 | \
 	mr12|\
 	mr16|\
 	zbt-we1526|\
--- master-fd30187-bak/target/linux/ar71xx/base-files/etc/board.d/01_leds	2018-03-13 11:01:52.327066070 -0400
+++ master-fd30187/target/linux/ar71xx/base-files/etc/board.d/01_leds	2018-03-13 15:53:38.925126185 -0400
@@ -445,7 +445,7 @@
 	ucidef_set_led_wlan "wlan" "WLAN" "hiwifi:blue:wlan-2p4" "phy0tpt"
 	;;
 hornet-ub|\
-hornet-ub-x2)
+gargoyle-pocket-router-v2)
 	ucidef_set_led_netdev "lan" "LAN" "alfa:blue:lan" "eth0"
 	ucidef_set_led_netdev "wan" "WAN" "alfa:blue:wan" "eth1"
 	ucidef_set_led_wlan "wlan" "WLAN" "alfa:blue:wlan" "phy0tpt"
--- master-fd30187-bak/target/linux/ar71xx/base-files/etc/diag.sh	2018-03-13 11:01:52.327066070 -0400
+++ master-fd30187/target/linux/ar71xx/base-files/etc/diag.sh	2018-03-13 15:54:32.765176105 -0400
@@ -222,7 +222,7 @@
 		status_led="hiwifi:blue:system"
 		;;
 	hornet-ub|\
-	hornet-ub-x2)
+	gargoyle-pocket-router-v2)
 		status_led="alfa:blue:wps"
 		;;
 	ja76pf|\
--- master-fd30187-bak/target/linux/ar71xx/image/legacy.mk	2018-03-13 11:01:52.331066147 -0400
+++ master-fd30187/target/linux/ar71xx/image/legacy.mk	2018-03-13 15:47:41.676715869 -0400
@@ -224,7 +224,11 @@
 alfa_ap120c_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,13312k(rootfs),1536k(kernel),1152k(unknown)ro,64k(art)ro;spi0.1:-(unknown)
 alfa_ap96_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,256k(u-boot-env)ro,13312k(rootfs),2048k(kernel),512k(caldata)ro,15360k@0x80000(firmware)
 alfa_mtdlayout_8M=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,6144k(rootfs),1600k(kernel),64k(nvram),64k(art)ro,7744k@0x50000(firmware)
-alfa_mtdlayout_16M=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,15936k(firmware),64k(nvram),64k(art)ro
+
+
+alfa_mtdlayout_16M=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,14336k(rootfs),1600k(kernel),64k(nvram),64k(art)ro,15936k@0x50000(firmware)
+tube2h_mtdlayout_16M=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,15936k(firmware),64k(nvram),64k(art)ro
+
 all0258n_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env),6272k(firmware),1536k(failsafe),64k(art)ro
 all0315n_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,256k(u-boot-env),13568k(firmware),2048k(failsafe),256k(art)ro
 ap96_mtdlayout=mtdparts=spi0.0:192k(u-boot)ro,64k(u-boot-env)ro,6144k(rootfs),1728k(kernel),64k(art)ro,7872k@0x40000(firmware)
@@ -894,8 +898,14 @@
 $(eval $(call SingleProfile,AthLzma,64k,BXU2000N2,bxu2000n-2-a1,BXU2000n-2-A1,ttyS0,115200,$$(bxu2000n2_mtdlayout),RKuImage))
 $(eval $(call SingleProfile,AthLzma,64k,CAP4200AG,cap4200ag,CAP4200AG,ttyS0,115200,$$(cap4200ag_mtdlayout),KRuImage))
 $(eval $(call SingleProfile,AthLzma,64k,DB120,db120,DB120,ttyS0,115200,$$(db120_mtdlayout),RKuImage))
-$(eval $(call SingleProfile,AthLzma,64k,HORNETUBx2,hornet-ub-x2,HORNET-UB,ttyATH0,115200,$$(alfa_mtdlayout_16M),KRuImage,65536))
-$(eval $(call SingleProfile,AthLzma,64k,TUBE2H16M,tube2h-16M,TUBE2H,ttyATH0,115200,$$(alfa_mtdlayout_16M),KRuImage,65536))
+
+
+#$(eval $(call SingleProfile,AthLzma,64k,HORNETUBx2,hornet-ub-x2,HORNET-UB,ttyATH0,115200,$$(alfa_mtdlayout_16M),KRuImage,65536))
+#$(eval $(call SingleProfile,AthLzma,64k,TUBE2H16M,tube2h-16M,TUBE2H,ttyATH0,115200,$$(alfa_mtdlayout_16M),KRuImage,65536))
+
+$(eval $(call SingleProfile,AthLzma,64k,GPRv2,gargoyle-pocket-router-v2,HORNET-UB,ttyATH0,115200,$$(alfa_mtdlayout_16M),RKuImage))
+$(eval $(call SingleProfile,AthLzma,64k,TUBE2H16M,tube2h-16M,TUBE2H,ttyATH0,115200,$$(tube2h_mtdlayout_16M),KRuImage,65536))
+
 
 $(eval $(call SingleProfile,Belkin,64k,F9K1115V2,f9k1115v2,F9K1115V2,ttyS0,115200,$$(f9k1115v2_mtdlayout),BR-6679BAC))
 
