--- a/target/linux/ar71xx/image/Makefile	2015-07-21 12:50:17.095881121 -0400
+++ b/target/linux/ar71xx/image/Makefile	2015-07-21 17:14:02.159863725 -0400
@@ -1171,7 +1171,8 @@
 	$(eval rootsize=$(call mtdpartsize,rootfs,$(4)))
 	$(call Sysupgrade/RKuImage,$(1),$(2),$(kernsize),$(rootsize))
 	if [ -e "$(call sysupname,$(1),$(2))" ]; then \
-		edimax_fw_header -m $(5) -v "OpenWrt$(REVISION)" \
+		edimax_header_name=$$$$( echo "OpenWrt$(REVISION)" | cut -b 1-10 ) \
+		edimax_fw_header -m $(5) -v "$$$$edimax_header_name" \
 			-n "uImage" \
 			-i $(KDIR_TMP)/vmlinux-$(2).uImage \
 			-o $(KDIR_TMP)/$(2)-uImage; \

