--- a/include/target.mk
+++ b/include/target.mk
@@ -91,9 +91,9 @@
 endif
 
 # include ujail on systems with enough storage
-ifeq ($(filter small_flash,$(FEATURES)),)
-  DEFAULT_PACKAGES+=procd-ujail
-endif
+#ifeq ($(filter small_flash,$(FEATURES)),)
+#  DEFAULT_PACKAGES+=procd-ujail
+#endif
 
 # Add device specific packages (here below to allow device type set from subtarget)
 DEFAULT_PACKAGES += $(DEFAULT_PACKAGES.$(DEVICE_TYPE))
