--- bhc_build_ar71xx-large/include/prereq-build.mk	2017-03-14 14:22:25.589623137 -0400
+++ bhc_build_ar71xx-extralarge-nand/include/prereq-build.mk	2017-09-06 00:01:19.152191786 -0400
@@ -145,7 +145,7 @@
 	svn --version | grep Subversion))
 
 $(eval $(call SetupHostCommand,git,Please install Git (git-core) >= 1.6.5, \
-	git clone 2>&1 | grep -- --recursive))
+	git clone 2>&1 | grep -- --recurs))
 
 $(eval $(call SetupHostCommand,file,Please install the 'file' package, \
 	file --version 2>&1 | grep file))
