--- a/tools/mtd-utils/Makefile	2015-07-21 12:50:17.047881924 -0400
+++ b/tools/mtd-utils/Makefile	2015-07-21 12:51:20.446821715 -0400
@@ -11,7 +11,7 @@
 
 PKG_SOURCE_VERSION:=92686f212c9a4e16891c6a3c57629cbf4f0f8360
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION)-$(PKG_SOURCE_VERSION).tar.gz
-PKG_SOURCE_URL:=git://git.infradead.org/mtd-utils.git
+PKG_SOURCE_URL:=git://github.com/ericpaulbishop/mtdutil.git
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_VERSION)
 PKG_MIRROR_MD5SUM:=154592f5bbb01e1e98bdd26c8a592305
