--- a/include/version.mk
+++ b/include/version.mk
@@ -30,7 +30,7 @@ VERSION_NICK:=$(call qstrip_escape,$(CON
 VERSION_NICK:=$(if $(VERSION_NICK),$(VERSION_NICK),$(RELEASE))
 
 VERSION_REPO:=$(call qstrip_escape,$(CONFIG_VERSION_REPO))
-VERSION_REPO:=$(if $(VERSION_REPO),$(VERSION_REPO),http://downloads.openwrt.org/%n/%v/%S/packages)
+VERSION_REPO:=$(if $(VERSION_REPO),$(VERSION_REPO),http://archive.openwrt.org/%n/%v/%S/packages)
 
 VERSION_DIST:=$(call qstrip_escape,$(CONFIG_VERSION_DIST))
 VERSION_DIST:=$(if $(VERSION_DIST),$(VERSION_DIST),OpenWrt)
