--- a/package/netifd/files/lib/netifd/dhcp.script	2012-08-27 02:10:27.877896974 -0400
+++ b/package/netifd/files/lib/netifd/dhcp.script	2012-09-01 16:00:04.590119277 -0400
@@ -35,6 +35,11 @@
 	done
 	proto_send_update "$INTERFACE"
 
+	
+	/sbin/uci -P /var/state set network.$INTERFACE.lease_server="$serverid" 
+	/sbin/uci -P /var/state set network.$INTERFACE.lease_lifetime="$lease" 
+	/sbin/uci -P /var/state set network.$INTERFACE.lease_acquired="$(sed -ne 's![^0-9].*$!!p' /proc/uptime)"
+
 	# TODO
 	# [ -n "$ntpsrv" ] && 	change_state network "$ifc" lease_ntpsrv "$ntpsrv"
 	# [ -n "$timesvr" ] && 	change_state network "$ifc" lease_timesrv "$timesvr"
