--- a/include/version.mk
+++ b/include/version.mk
@@ -16,5 +16,5 @@
 
 VERSION_NUMBER:=$(call qstrip,$(CONFIG_VERSION_NUMBER))
-VERSION_NUMBER:=$(if $(VERSION_NUMBER),$(VERSION_NUMBER),12.09-rc1)
+VERSION_NUMBER:=$(if $(VERSION_NUMBER),$(VERSION_NUMBER),12.09-beta2)
 
 VERSION_CODE:=$(call qstrip,$(CONFIG_VERSION_NUMBER))
@@ -25,5 +25,5 @@
 
 VERSION_REPO:=$(call qstrip,$(CONFIG_VERSION_REPO))
-VERSION_REPO:=$(if $(VERSION_REPO),$(VERSION_REPO),http://downloads.openwrt.org/attitude_adjustment/12.09-rc1/%S/packages)
+VERSION_REPO:=$(if $(VERSION_REPO),$(VERSION_REPO),http://downloads.openwrt.org/attitude_adjustment/12.09-beta2/%S/packages)
 
 VERSION_DIST:=$(call qstrip,$(CONFIG_VERSION_DIST))
--- a/package/base-files/image-config.in
+++ b/package/base-files/image-config.in
@@ -170,5 +170,5 @@
 		string
 		prompt "Release repository" if VERSIONOPT
-		default "http://downloads.openwrt.org/attitude_adjustment/12.09-rc1/%S/packages" if VERSIONOPT
+		default "http://downloads.openwrt.org/attitude_adjustment/12.09-beta2/%S/packages" if VERSIONOPT
 		help
 			This is the repository address embedded in the image, it defaults
