--- a/target/linux/ar71xx/base-files/etc/uci-defaults/01_leds	2015-08-17 15:05:37.786082500 +1000
+++ b/target/linux/ar71xx/base-files/etc/uci-defaults/01_leds	2015-08-19 13:59:48.544426900 +1000
@@ -156,6 +156,10 @@
 	ucidef_set_led_wlan "wlan2g" "WLAN 2.4 GHz" "d-link:blue:wlan2g" "phy0tpt"
 	;;
 
+dir-835-a1)
+	ucidef_set_led_netdev "wav" "WAN" "d-link:amber:planet" "eth0.2"
+	;;
+
 dlan-pro-500-wp)
 	ucidef_set_led_default "power" "System Power" "devolo:green:status" "1"
 	ucidef_set_led_netdev "lan" "Ethernet Activity" "devolo:green:eth" "br-lan"
