--- a/target/linux/ar71xx/base-files/etc/uci-defaults/routerstation-pro	(wersja 0)
+++ b/target/linux/ar71xx/base-files/etc/uci-defaults/routerstation-pro	(wersja 0)
@@ -0,0 +1,24 @@
+#!/bin/sh
+#
+# Copyright (C) 2010 OpenWrt.org
+#
+
+. /lib/ar71xx.sh
+
+board=$(ar71xx_board_name)
+
+rspro_set_wlan_led() {
+	uci batch <<EOF
+set system.wlan_led=led
+set system.wlan_led.name=WLAN
+set system.wlan_led.sysfs=ubnt:green:rf
+set system.wlan_led.dev=wlan0
+set system.wlan_led.trigger=netdev
+set system.wlan_led.mode="link rx tx"
+commit system
+EOF
+}
+
+if [ "${board}" == "routerstation-pro" ]; then
+	rspro_set_wlan_led
+fi
