--- a/target/linux/generic/files/drivers/net/phy/mvsw61xx.c
+++ b/target/linux/generic/files/drivers/net/phy/mvsw61xx.c
@@ -754,6 +754,9 @@ static int mvsw61xx_probe(struct platfor
 	case MV_IDENT_VALUE_6176:
 		model_str = MV_IDENT_STR_6176;
 		break;
+	case MV_IDENT_VALUE_6352:
+		model_str = MV_IDENT_STR_6352;
+		break;
 	default:
 		dev_err(&pdev->dev, "No compatible switch found at 0x%02x\n",
 				state->base_addr);
@@ -817,6 +820,7 @@ static const struct of_device_id mvsw61x
 	{ .compatible = "marvell,88e6171" },
 	{ .compatible = "marvell,88e6172" },
 	{ .compatible = "marvell,88e6176" },
+	{ .compatible = "marvell,88e6352" },
 	{ }
 };
 MODULE_DEVICE_TABLE(of, mvsw61xx_match);
--- a/target/linux/generic/files/drivers/net/phy/mvsw61xx.h
+++ b/target/linux/generic/files/drivers/net/phy/mvsw61xx.h
@@ -222,6 +222,9 @@ enum {
 #define MV_IDENT_VALUE_6176		0x1760
 #define MV_IDENT_STR_6176		"MV88E6176"
 
+#define MV_IDENT_VALUE_6352		0x3520
+#define MV_IDENT_STR_6352		"MV88E6352"
+
 #define MV_PVID_MASK			0x0fff
 
 struct mvsw61xx_state {
