--- backfire-src/package/base-files/files/lib/network/config.sh	2010-08-25 21:30:09.660614150 -0400
+++ ar71xx-src/package/base-files/files/lib/network/config.sh	2010-08-28 13:59:43.251631569 -0400
@@ -152,13 +152,20 @@
 		}
 	}
 
-	# Setup VLAN interfaces
-	add_vlan "$iface" && return 1
-	ifconfig "$iface" 2>/dev/null >/dev/null || return 0
-
-	# Setup bridging
+	
 	local iftype
 	config_get iftype "$config" type
+
+
+	if [ "$iftype" = "bridge" ] && [ -z "$iface" -o "$iface" = "none" ] ; then
+		echo "configuring empty bridge"
+	else
+		# Setup VLAN interfaces
+		add_vlan "$iface" && return 1
+		ifconfig "$iface" 2>/dev/null >/dev/null || return 0
+	fi
+	
+	# Setup bridging
 	case "$iftype" in
 		bridge)
 			local macaddr
--- backfire-src/package/base-files/files/sbin/ifup	2010-08-25 21:30:09.760613809 -0400
+++ ar71xx-src/package/base-files/files/sbin/ifup	2010-08-28 13:50:19.912881847 -0400
@@ -18,6 +18,7 @@
 scan_interfaces
 
 config_get ifname "$1" device
+if [ -z "$ifname" ] || [ "$ifname" = "br-$1" ] ; then ifname="none" ; fi
 for dev in ${ifname:-$1}; do
 	setup_interface "$dev" "$1"
 done
--- backfire-src/package/base-files/files/etc/init.d/network	2010-08-25 21:30:09.729363604 -0400
+++ ar71xx-src/package/base-files/files/etc/init.d/network	2010-08-28 13:51:24.001631458 -0400
@@ -9,6 +9,18 @@
 
 	include /lib/network
 	setup_switch
+	
+	#bring up bridge interfaces with no iface or iface set to "none"
+	networks=$( uci show network | grep "=interface$" | sed 's/^network.//g' | sed 's/=interface//g' )
+	for n in $networks ; do
+		nt=$(uci get network.$n.type 2>/dev/null)
+		ni=$(uci get network.$n.ifname 2>/dev/null)
+		if [ "$nt" = "bridge" ] && [ -z "$ni" -o "$ni" = "none" ] ; then
+			ifup $n
+		fi
+	done
+	
+	
 	grep -qs config /etc/config/wireless && {
 		/sbin/wifi up
 	} || {
