--- a/target/linux/ar71xx/image/Makefile	2014-11-30 11:21:41.870644632 -0500
+++ b/target/linux/ar71xx/image/Makefile	2014-12-02 07:17:24.058812497 -0500
@@ -574,7 +578,8 @@
 	-$(STAGING_DIR_HOST)/bin/mkfwimage2 -f 0x9f000000 \
 		-v $(5).$(6).v6.0.0-OpenWrt-$(REVISION) \
 		-p jffs2:0x50000:0xf60000:0:0:$(call sysupname,$(1),$(2)) \
-		-o $(call factoryname,$(1),$(2))
+		-o $(call factoryname,$(1),$(2)) \
+	echo ""
 endef
 
 # $(4) = board (XS2, XS5, RS, XM)
@@ -591,7 +596,8 @@
 			-k $(KDIR_TMP)/$(2)-mtdpart-kernel.bin \
 			-r $(KDIR_TMP)/$(2)-mtdpart-rootfs.bin \
 			-o $(call factoryname,$(1),$(2)); \
-	fi
+	fi ; \
+	echo ""
 endef
 
 Image/Build/UBNTXM/initramfs=$(call MkuImageLzma/initramfs,$(2),$(3) $(ubntxm_mtdlayout))
