--- a/package/network/services/relayd/files/relay.hotplug	2016-03-15 22:17:41.922342700 +1100
+++ b/package/network/services/relayd/files/relay.hotplug	2016-03-15 22:22:46.192371400 +1100
@@ -1,2 +1,13 @@
 #!/bin/sh
+#enable access from client network
+[ "$INTERFACE" = wwan ] || exit 0
+[ "$ACTION" = ifup -o "$ACTION" = ifupdate ] || exit 0
+
+. /lib/functions/network.sh; network_get_ipaddr ip wwan;
+
+uci set network.bridgecfg.ipaddr=$ip
+uci commit network
+logger -t relayd "Assigning $ip to bridge interface"
+
 /etc/init.d/relayd enabled && /etc/init.d/relayd start
+logger -t relayd "Reloading relayd due to $ACTION of $INTERFACE ($DEVICE)"
