--- backfire-src/package/base-files/files/usr/share/udhcpc/default.script	2010-09-06 20:59:01.464875249 -0400
+++ ar71xx-src/package/base-files/files/usr/share/udhcpc/default.script	2010-09-11 11:50:28.610783288 -0400
@@ -161,6 +161,7 @@
 	# UCI State
 	change_state network "$ifc" lease_server "$serverid"
 	change_state network "$ifc" lease_acquired "$(date '+%s')"
+	change_state network "$ifc" lease_acquired_uptime "$(cat /proc/uptime | sed 's/\..*$//g' | sed 's/ .*$//g')"
 	change_state network "$ifc" lease_lifetime "$lease"
 	[ -n "$ntpsrv" ] && 	change_state network "$ifc" lease_ntpsrv "$ntpsrv"
 	[ -n "$timesvr" ] && 	change_state network "$ifc" lease_timesrv "$timesvr"
