--- backfire-a/package/hostapd/files/wpa_supplicant.sh	2011-11-24 22:56:18.374901091 -0500
+++ backfire-b/package/hostapd/files/wpa_supplicant.sh	2011-11-25 13:46:44.919294338 -0500
@@ -27,10 +27,6 @@
 	config_get mode "$vif" mode
 	config_get ifname "$vif" ifname
 	config_get_bool wds "$vif" wds 0
-	[ -z "$bridge" ] || [ "$mode" = ap ] || [ "$mode" = sta -a $wds -eq 1 ] || {
-		echo "wpa_supplicant_setup_vif($ifname): Refusing to bridge $mode mode interface"
-		return 1
-	}
 
 	case "$enc" in
 		*none*)
