--- a/target/imagebuilder/files/Makefile	2014-11-08 12:48:51.636428286 -0500
+++ b/target/imagebuilder/files/Makefile	2014-11-30 11:18:17.757952447 -0500
@@ -151,6 +151,8 @@
 	)
 	rm -f $(TARGET_DIR)/usr/lib/opkg/info/*.postinst
 	$(if $(CONFIG_CLEAN_IPKG),rm -rf $(TARGET_DIR)/usr/lib/opkg)
+	$(if $(CONFIG_CLEAN_IPKG_PARTIAL),rm -rf $(TARGET_DIR)/usr/lib/opkg/info/*.postinst)
+	$(if $(CONFIG_CLEAN_IPKG_PARTIAL),rm -rf $(TARGET_DIR)/usr/lib/opkg/info/*.control)
 
 build_image: FORCE
 	@echo
--- a/package/Makefile	2014-11-08 12:48:51.636428286 -0500
+++ b/package/Makefile	2014-11-30 11:18:17.757952447 -0500
@@ -118,6 +118,8 @@
 	@-find $(TARGET_DIR) -name '.#*' | $(XARGS) rm -f
 	rm -f $(TARGET_DIR)/usr/lib/opkg/info/*.postinst
 	$(if $(CONFIG_CLEAN_IPKG),rm -rf $(TARGET_DIR)/usr/lib/opkg)
+	$(if $(CONFIG_CLEAN_IPKG_PARTIAL),rm -rf $(TARGET_DIR)/usr/lib/opkg/info/*.postinst)
+	$(if $(CONFIG_CLEAN_IPKG_PARTIAL),rm -rf $(TARGET_DIR)/usr/lib/opkg/info/*.control)
 	$(call mklibs)
 
 PASSOPT=""
