--- a/target/linux/ar71xx/generic/profiles/alfa.mk	2014-11-08 12:48:51.624428482 -0500
+++ b/target/linux/ar71xx/generic/profiles/alfa.mk	2014-11-24 19:43:16.170793391 -0500
@@ -30,17 +30,19 @@
 $(eval $(call Profile,HORNETUB))
 
 
-define Profile/HORNETUBx2
-	NAME:=ALFA Network Hornet-UB-x2 board (16MB flash, 64MB ram)
-	PACKAGES:=kmod-usb-core kmod-usb2 kmod-ledtrig-usbdev
-endef
-
-define Profile/HORNETUBx2/Description
-	Package set optimized for the ALFA Network Hornet-UB board with 16MB
-	flash and 64MB ram.
-endef
 
-$(eval $(call Profile,HORNETUBx2))
+define Profile/GPRv2 
+	NAME:=Gargoyle Pocket Router v2
+	PACKAGES:=kmod-usb-core kmod-usb2 kmod-ledtrig-usbdev 
+endef 
+	 
+define Profile/GPRv2/Description 
+	Package set optimized for Gargoyle Pocket Router v2
+	Based on ALFA Network Hornet-UB board with 16MB flash and 64MB ram. 
+endef 
+
+$(eval $(call Profile,GPRv2))
+
 
 
 define Profile/ALFANX
--- a/target/linux/ar71xx/image/Makefile	2014-11-08 12:48:51.628428417 -0500
+++ b/target/linux/ar71xx/image/Makefile	2014-11-24 19:40:29.785430697 -0500
@@ -239,7 +239,11 @@
 
 alfa_ap96_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,256k(u-boot-env)ro,13312k(rootfs),2048k(kernel),512k(caldata)ro,15360k@0x80000(firmware)
 alfa_mtdlayout_8M=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,6144k(rootfs),1600k(kernel),64k(nvram),64k(art)ro,7744k@0x50000(firmware)
-alfa_mtdlayout_16M=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,15936k(firmware),64k(nvram),64k(art)ro
+
+tube2h_mtdlayout_16M=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,15936k(firmware),64k(nvram),64k(art)ro
+alfa_mtdlayout_16M=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,14336k(rootfs),1600k(kernel),64k(nvram),64k(art)ro,15936k@0x50000(firmware)
+
+
 all0258n_mtdlayout=mtdparts=spi0.0:256k(u-boot),64k(u-boot-env),6272k(firmware),1536k(failsafe),64k(art)
 all0315n_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,256k(u-boot-env),13568k(firmware),2048k(failsafe),256k(art)ro
 ap81_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,5120k(rootfs),2688k(kernel),64k(art)ro,7808k@0x50000(firmware)
@@ -1079,9 +1083,14 @@
 $(eval $(call SingleProfile,AthLzma,64k,EWDORINAP,ew-dorin,EW-DORIN,ttyATH0,115200,$$(ew-dorin_mtdlayout_4M),KRuImage,65536))
 $(eval $(call SingleProfile,AthLzma,64k,EWDORINRT,ew-dorin-router,EW-DORIN-ROUTER,ttyATH0,115200,$$(ew-dorin_mtdlayout_4M),KRuImage,65536))
 $(eval $(call SingleProfile,AthLzma,64k,EWDORIN16M,ew-dorin-16M,EW-DORIN,ttyATH0,115200,$$(ew-dorin_mtdlayout_16M),KRuImage,65536))
-$(eval $(call SingleProfile,AthLzma,64k,HORNETUBx2,hornet-ub-x2,HORNET-UB,ttyATH0,115200,$$(alfa_mtdlayout_16M),KRuImage,65536))
+
+
+#$(eval $(call SingleProfile,AthLzma,64k,HORNETUBx2,hornet-ub-x2,HORNET-UB,ttyATH0,115200,$$(alfa_mtdlayout_16M),KRuImage,65536))
+$(eval $(call SingleProfile,AthLzma,64k,GPRv2,gargoyle-pocket-router-v2,HORNET-UB,ttyATH0,115200,$$(alfa_mtdlayout_16M),RKuImage))
+
+
 $(eval $(call SingleProfile,AthLzma,64k,PB92,pb92,PB92,ttyS0,115200,$$(pb92_mtdlayout),KRuImage))
-$(eval $(call SingleProfile,AthLzma,64k,TUBE2H16M,tube2h-16M,TUBE2H,ttyATH0,115200,$$(alfa_mtdlayout_16M),KRuImage,65536))
+$(eval $(call SingleProfile,AthLzma,64k,TUBE2H16M,tube2h-16M,TUBE2H,ttyATH0,115200,$$(tube2h_mtdlayout_16M),KRuImage,65536))
 $(eval $(call SingleProfile,AthLzma,64k,WLR8100,wlr8100,WLR8100,ttyS0,115200,$$(wlr8100_mtdlayout),KRuImage))
 
 $(eval $(call SingleProfile,CameoAP91,64kraw,DIR600A1,dir-600-a1,DIR-600-A1,ttyS0,115200,"AP91-AR7240-RT-090223-00"))
--- a/target/linux/ar71xx/base-files/lib/ar71xx.sh	2014-11-08 12:48:51.628428417 -0500
+++ b/target/linux/ar71xx/base-files/lib/ar71xx.sh	2014-11-24 19:44:26.633676532 -0500
@@ -370,7 +370,7 @@
 		fi
 
 		if [ "x$size" = "x00f90000" ]; then
-			name="hornet-ub-x2"
+			name="gargoyle-pocket-router-v2"
 		fi
 		;;
 	*LS-SR71)
--- a/target/linux/ar71xx/base-files/lib/upgrade/platform.sh	2014-11-08 12:48:51.628428417 -0500
+++ b/target/linux/ar71xx/base-files/lib/upgrade/platform.sh	2014-11-24 19:46:00.188193682 -0500
@@ -155,6 +155,7 @@
 	ap96 | \
 	db120 | \
 	hornet-ub | \
+	gargoyle-pocket-router-v2 | \
 	bxu2000n-2-a1 | \
 	zcn-1523h-2 | \
 	zcn-1523h-5)
--- a/target/linux/ar71xx/base-files/etc/uci-defaults/01_leds	2014-11-08 12:48:51.628428417 -0500
+++ b/target/linux/ar71xx/base-files/etc/uci-defaults/01_leds	2014-11-24 19:47:43.122562182 -0500
@@ -129,6 +129,7 @@
 	ucidef_set_led_wlan "wlan" "WLAN" "hiwifi:blue:wlan-2p4" "phy0tpt"
 	;;
 
+gargoyle-pocket-router-v2|\
 hornet-ub)
 	ucidef_set_led_netdev "lan" "LAN" "alfa:blue:lan" "eth0"
 	ucidef_set_led_netdev "wan" "WAN" "alfa:blue:wan" "eth1"
