--- atheros-src/target/linux/ar71xx/generic/profiles/alfa.mk	2013-10-16 23:01:31.192952467 -0400
+++ ar71xx-src/target/linux/ar71xx/generic/profiles/alfa.mk	2013-11-02 21:09:16.787014317 -0400
@@ -40,3 +40,16 @@
 
 $(eval $(call Profile,ALFANX))
 
+
+define Profile/GPRv2 
+	NAME:=Gargoyle Pocket Router v2
+	PACKAGES:=kmod-usb-core kmod-usb2 kmod-ledtrig-usbdev 
+endef 
+	 
+define Profile/GPRv2/Description 
+	Gargoyle Pocket Router v2
+	Based on ALFA Network Hornet-UB board with 16MB flash and 64MB ram. 
+endef 
+
+$(eval $(call Profile,GPRv2))
+
--- atheros-src/target/linux/ar71xx/image/Makefile	2013-10-16 23:01:31.496947263 -0400
+++ ar71xx-src/target/linux/ar71xx/image/Makefile	2013-11-03 01:45:28.392758625 -0400
@@ -186,6 +186,9 @@
 wndr3700v2_mtdlayout=mtdparts=spi0.0:320k(u-boot)ro,128k(u-boot-env)ro,1024k(kernel),14848k(rootfs),64k(art)ro,15872k@0x70000(firmware)
 zcn1523h_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,6208k(rootfs),1472k(kernel),64k(configure)ro,64k(mfg)ro,64k(art)ro,7680k@0x50000(firmware)
 
+
+alfa_mtdlayout_16M=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,14336k(rootfs),1600k(kernel),64k(nvram),64k(art)ro,15936k@0x50000(firmware)
+
 define Image/BuildKernel
 	cp $(KDIR)/vmlinux.elf $(VMLINUX).elf
 	cp $(KDIR)/vmlinux $(VMLINUX).bin
@@ -824,6 +827,9 @@
 $(eval $(call SingleProfile,AthLzma,$(fs_64k),EWDORINRT,ew-dorin-router,EW-DORIN-ROUTER,ttyATH0,115200,$$(ew-dorin_mtdlayout_4M),1048576,2752512,KRuImage))
 $(eval $(call SingleProfile,AthLzma,$(fs_64k),PB92,pb92,PB92,ttyS0,115200,$$(pb92_mtdlayout),917504,2818048,KRuImage))
 
+
+$(eval $(call SingleProfile,AthLzma,$(fs_64k),GPRv2,gargoyle-pocket-router-v2,HORNET-UB,ttyATH0,115200,$$(alfa_mtdlayout_16M),1638400,14680064,RKuImage))
+
 $(eval $(call SingleProfile,Cameo7240,$(fs_64k),DIR600A1,dir-600-a1,DIR-600-A1,ttyS0,115200,"AP91-AR7240-RT-090223-00"))
 $(eval $(call SingleProfile,Cameo7240,$(fs_64k),DIR601A1,dir-601-a1,DIR-600-A1,ttyS0,115200,"AP91-AR7240-RT-090223-02"))
 $(eval $(call SingleProfile,Cameo7240,$(fs_64k),DIR615E4,dir-615-e4,DIR-615-E4,ttyS0,115200,"AP99-AR7240-RT-091105-05"))
--- atheros-src/target/linux/ar71xx/base-files/lib/ar71xx.sh	2013-10-16 23:01:31.492947330 -0400
+++ ar71xx-src/target/linux/ar71xx/base-files/lib/ar71xx.sh	2013-11-03 03:10:40.895180929 -0500
@@ -229,7 +229,16 @@
 		name="jwap003"
 		;;
 	*"Hornet-UB")
-		name="hornet-ub"
+		local size 
+		size=$(awk '/firmware/ { print $2 }' /proc/mtd) 
+
+		if [ "x$size" = "x00790000" ]; then 
+			name="hornet-ub" 
+		fi 
+
+		if [ "x$size" = "x00f90000" ]; then 
+			name="gargoyle-pocket-router-v2" 
+		fi 
 		;;
 	*LS-SR71)
 		name="ls-sr71"
--- atheros-src/target/linux/ar71xx/base-files/lib/upgrade/platform.sh	2013-10-16 23:01:31.492947330 -0400
+++ ar71xx-src/target/linux/ar71xx/base-files/lib/upgrade/platform.sh	2013-11-04 01:21:44.642080525 -0500
@@ -92,6 +92,7 @@
 	ap96 | \
 	db120 | \
 	hornet-ub | \
+	gargoyle-pocket-router-v2 | \
 	zcn-1523h-2 | \
 	zcn-1523h-5)
 		[ "$magic_long" != "68737173" -a "$magic_long" != "19852003" ] && {
--- a/target/linux/ar71xx/base-files/etc/uci-defaults/leds	2014-01-16 21:23:39.556582430 -0500
+++ b/target/linux/ar71xx/base-files/etc/uci-defaults/leds	2014-02-17 17:25:19.933304852 -0500
@@ -69,6 +69,7 @@
 	ucidef_set_led_usbdev "usb" "USB" "d-link:blue:usb" "1-1"
 	;;
 
+gargoyle-pocket-router-v2|\
 hornet-ub)
 	ucidef_set_led_netdev "lan" "LAN" "alfa:blue:lan" "eth0"
 	ucidef_set_led_netdev "wan" "WAN" "alfa:blue:wan" "eth1"
